### DISCUSSION TITLE
Mouse  and keyboard events (on steroids)

### DISCUSSION THREAD

**Discussion by DaMilyutin** [slint-ui/slint#790] [Ideas]
I know there was some work done for mouse events.
However, I'd like to share some thoughts, discuss it and clarify how it is done.

I'd categorize events in groups
1. Move (no buttons pressed or clicked) => outputs position shift
2. Drag (button pressed and held)  => outputs position shift and buttons pressed (may be combination of few keyboard and mouse buttons)
3. Click (button pressed and released once or more) -> outputs buttons clicked (may be combination of few KB and mouse buttins) shifts are meant to be zero.
4. Mouse wheel up/down. wheel may be shifted with different "force", so shift value is like extra dimension.
 
As for C++ developer it seems to me that bitmask will be best option to handle pressed buttons.
Maybe on can have separated keyboard and mouse events somehow and then fuse them.
From other hand one can just have InputEvent and process it as he desires.
All KB keys will probably fit in 256 bits and mouse keys in 8 bits.
As for clicks it actually harder task. One need to write sort of driver to interpret mouse inputs,
I'd add field fot number of clicks. 0 - pressed, 1 - clicked, 2 - double clicked, etc.

In C++I'd model it like
```
struct MouseEvent
{
     int dx, dy;           // position changed
     uint8_t buttons; // if 0 nothing pressed
     uint8_t clicks;     // pressed, single, double, ...
     int8_t wheel;      // up, down shift; 0 if no shift  
};

struct KeyboardEvent
{
    uint64_t buttons[4]; // naive bitmask for buttons (ex. English)
};
```

This approach will allow for user of sixtyfps to process events with many keys pressed. 
this includes events with modifier keys like ex. shift+click.

So is such approach interesting for you?
Where can I see usage of current iterface?

**Comment by ogoffart** [comment]
Apart from the wheel event, i believe the current API covers already what you suggest.

What use-case do you have in mind?

**Reply by DaMilyutin** [reply]
I'm going to use it to control 2d/3d plot. 
Like pan (L-mouse drag), zoom (M-Wheel), rotate 3d (R-mouse drag).

I wanted to add horizontal and vertical pan/zoom modifiers with key.H+drag, key.V+drag.

I actually thought more and there is one more category to add. 
5. Mouse appeared over the area and mouse left the area.
On these events one can change mouse cursor. I've seen such request in discussions or issues.

**Reply by DaMilyutin** [reply]
@ogoffart , @tronical, are there any updates on input events? 
I mean mouse wheel.

Is there a way to get maybe all "native" events and to process them on my part somehow?
Or have access to some other low-level backend?

**Reply by ogoffart** [reply]
The issue is https://github.com/slint-ui/slint/issues/1280

I don't think there was progress.  

We should just decide on the API. The implementation should be fairly trivial

**Reply by ogoffart** [reply]
There was also this discussion recently https://github.com/slint-ui/slint/discussions/2182

**Comment by newtome8888** [comment]
> Apart from the wheel event, i believe the current API covers already what you suggest.
> 
> What use-case do you have in mind?


Is there any example for how to handle mouse move in area and mouse leave area? I relly need to handle these events.

**Reply by tronical** [reply]
In the master branch, pointer-event is now also called for move events: https://github.com/slint-ui/slint/pull/4078

we don’t have an event for leave/enter yet. My feeling is that this would make sense.
